### PR TITLE
Swap DrupalVM for ddev.

### LINKED ISSUE #3641: Move on from Drupal VM?
There is a conversation going on about the future of Drupal VM: https://github.com/geerlingguy/drupal-vm/issues/2164

It doesn't seem, based on the conversation so far, that there is a clear path forward for Drupal VM. Jeff Geerling has committed to only minimal maintenance until he can't handle that anymore and then he's going to deprecate it.

While moving to a different virtualization paradigm is not a small thing, it is worth noting that we collectively spent a lot of hours resolving issues related to our current solution and figuring out how to continue using it. The introduction of the M1 Macs also raises questions about whether this is still the right solution for us.

Lando is an obvious alternative both because we have used it before and I think that docker has already figured out M1 compatibility.

Just wanted to have a place to keep the discussion rolling.

## Proposed Solution

- [x] Survey the options that provide drupal/acquia blt integration for M1 macs - https://github.com/uiowa/uiowa/issues/3641#issuecomment-970403213

Not that Slack support is key, but the current Drupal slack community numbers are ddev (801), lando (464), docksal (166). Releases also fall in that order ddev (146), lando (141), docksal (45). All seem to be releasing at a pretty regular schedule.

Lando pulls ahead in collaborators (193), ddev (159), docksal (62).

## Next Steps

- [x] Establish an experimental PR for ddev - https://github.com/uiowa/uiowa/pull/4458
- [x] Establish an experimental PR for lando - https://github.com/uiowa/uiowa/pull/4470
- [x] Try docksal
- [x] Test solutions on older/newer (M1) macs
- [x] Collectively decide on path forward?

### PR DESCRIPTION
**Pull request by joewhitsitt**
Resolves https://github.com/uiowa/uiowa/issues/3641

# DDEV UIowa Setup

_Note: If you don't have a current setup (M1), you will have to follow some onboarding docs to get git and homebrew setup._

- `vagrant halt` on your **current** setup
- [install docker 4.2.0 for mac](https://docs.docker.com/desktop/mac/release-notes/#docker-desktop-420) - _there is a bug with 4.3.0 and [here is the workaround if you find yourself up that creek](https://github.com/drud/ddev/issues/3404)_
- open the docker app and make sure it is "running"
- brew install drud/ddev/ddev
- mkcert -install
- git checkout `ddev` or `main` if this has been merged (hello future)
- if you were testing ddev elsewhere, headover there and `ddev stop --unlist uiowa`
- ddev auth ssh
- ddev composer install
- ddev start
- ddev auth ssh
- ddev ssh
- blt frontend
- blt bis
- blt ds

## Additional Notes

- the new local domain format is *.uiowa.ddev.site, so hr.uiowa.edu is hr.uiowa.ddev.site and gis.sites.uiowa.edu is sitesgis.uiowa.ddev.site
- If you have an existing docker install, make sure it is up to date, but not too far. 4.2.0
- ~~This does not clean up the drupalvm stuff completely.~~ 039c7d4
- You might have to run `ddev mutagen sync` if your containers seem out of sync.
- All Git operations need to happen on the host. BLT commands involving git were updated to require being run on host.
- Checkout the docs for more tidbits: https://ddev.readthedocs.io/en/stable/ (e.g. toggle xdebug, etc.)

## To Test

Try to perform normal local environment tasks:

- `blt frontend`
- `blt ds --site=sandbox.uiowa.edu`
- `drush @sandbox.local cr`

Additional resources:
https://iowaweb.slack.com/archives/C8SM82ZFZ/p1638808178159000

### CODE REVIEW

**Comment by richardbporter** [comment]
Checking this out since my local is broken. I might try to get it working in place instead of cloning fresh.

**Comment by joewhitsitt** [comment]
You might have to remove those local.settings and `blt bis` if you are trying this in your existing setup

**Comment by richardbporter** [comment]
I skipped this part b/c it seems tailored to single-site:
```
create a config.web_environment.yaml file in .ddev and add your acquia auth key/secret there using the following format. This is git ignored.
web_environment:
- ACQUIA_API_KEY=asdf
- ACQUIA_API_SECRET=asdf
ddev pull acquia
```

but I'm already up and running `dsa`. 👍🏻

**Comment by richardbporter** [comment]
Should we be ignoring everything in the .ddev directory except certain files? I'm seeing changes and untracked files there. Unclear from the docs here: https://ddev.readthedocs.io/en/stable/users/topics/whats_in_ddev_dir/

**Comment by joewhitsitt** [comment]
A fresh git clone/pull this is all it starts with, so i am not sure i understand. If the single stuff is not necessary for us, just the config file is all that is needed.

Screen Shot 2021-11-19 at 3 57 45 PM

**Comment by richardbporter** [comment]
I was seeing this when running a `git status` after starting up: 
```
Untracked files:
  (use "git add <file>..." to include in what will be committed)
        .ddev/.dbimageBuild/
        .ddev/.ddev-docker-compose-base.yaml
        .ddev/.ddev-docker-compose-full.yaml
        .ddev/.global_commands/
        .ddev/.homeadditions/
        .ddev/.webimageBuild/
        .ddev/apache/
        .ddev/commands/
        .ddev/homeadditions/
        .ddev/mutagen/
        .ddev/nginx_full/
        .ddev/providers/README.txt
        .ddev/providers/acquia.yaml.example
        .ddev/providers/localfile.yaml.example
        .ddev/providers/pantheon.yaml.example
        .ddev/providers/platform.yaml.example
        .ddev/providers/rsync.yaml.example

```

Seeing how quickly I can get #4470 working... :)

**Comment by joewhitsitt** [comment]
do you not have a .ddev/.gitignore file?

**Comment by richardbporter** [comment]
> do you not have a .ddev/.gitignore file?

I do now. Swear that wasn't there before...

This PR is leaning towards "just working" as opposed to #4470. Curious if @e-marie-w might have time to try.

**Comment by joewhitsitt** [comment]
My understanding is that Marie is working through the steps as I type this. Not sure at what point the PR was pulled. Might not have the latest commits

**Comment by joewhitsitt** [comment]
Our package.json might need updating to handle M1. Seeing all kinds of issues related to node-sass and python for M1s. We are running node:14 due to node-sass (https://github.com/sass/node-sass/issues/3033) and nodejs supports M1 starting at node:15.

Temp workaround would be to set the terminal window to use Rosetta https://github.com/nodejs/build/issues/2474

I don't know if it is this simple (when starting a new terminal window):
`arch -x86_64 zsh`

Long term, we need to adopt a higher node version and make sure our build works with the update dependencies.

**Comment by richardbporter** [comment]
We should probably update to Node 16: https://nodejs.org/en/about/releases/

File: docroot/sites/default/settings/default.local.settings.php [hunk 1]
```diff
@@ -7,6 +7,16 @@
...
+$host = '${drupal.db.host}';
+$port = '${drupal.db.port}';
+
+// If DDEV_PHP_VERSION is not set but IS_DDEV_PROJECT *is*, it means we're running (drush) on the host,
+// so use the host-side bind port on docker IP
+if (empty(getenv('DDEV_PHP_VERSION') && getenv('IS_DDEV_PROJECT') == 'true')) {
+  $host = "127.0.0.1";
+  $port = 53317;
+}
+
```
Comment (richardbporter) [inline docroot/sites/default/settings/default.local.settings.php]: Where did this snippet come from again?

Comment (richardbporter) [reply docroot/sites/default/settings/default.local.settings.php]: I removed this as it seems to allow running Drush on your host which is cool but doesn't really work with multisite Drush aliases. It also breaks Travis (I think.).

**Comment by richardbporter** [comment]
`ddev tableplus` 👍🏻

File: docroot/sites/default/settings/default.local.settings.php (hunk 1 above)
Comment (richardbporter) [reply docroot/sites/default/settings/default.local.settings.php]: Maybe not...

Comment (richardbporter) [reply docroot/sites/default/settings/default.local.settings.php]: I think we can leave this out.

Comment (joewhitsitt) [reply docroot/sites/default/settings/default.local.settings.php]: a generic ddev setup creates it, but then i saw somewhere that multisite shouldn't use ddev defaults for this and instead copy out and paste it for use elsewhere. we have `disable_settings_management` turned on

**Comment by richardbporter** [comment]
Having trouble getting Behat tests running locally. Drupal tests seem to be working.

**Comment by richardbporter** [comment]
I was able to get `blt frontend` working with 92d4382. I tried Node 15 and 16 but got all kinds of errors.

**Comment by e-marie-w** [comment]
`blt frontend` now works on the M1! Default and sandbox sites are displaying correctly! Thank you @richardbporter!!

**Comment by richardbporter** [comment]
> Having trouble getting Behat tests running locally. Drupal tests seem to be working.

Still having trouble with this. I imagine we'll have the same problem with #4470 but I haven't tested that. 

It might be a good time to remove Behat altogether - here are some reasons:

- We currently only have three, minimal Behat tests. :)
- With #4467, the BLT Behat integration was moved out of BLT core and into a standalone, community maintained plugin. :(
- Behat setup and configuration is confusing. 
- We've had issues with random failures in the past: https://github.com/uiowa/uiowa/issues/2203
- Behat session management is difficult with uiowa_auth enabled.
- Percy has proven fairly useful with minimal effort.
- Functional JS tests can do about the same thing and don't require the extra steps of using Gherkin syntax and custom feature contexts.

I still think Behat is ideal for testing our platform as a whole (profile + config) - it would be perfect for #4000. It's also probably more performant that functional JS tests since those run in isolation and would need to reinstall sitenow for every test. Still, this is all theoretical and not worth much without actual working tests. Maybe the POC Kernel tests in https://github.com/uiowa/uiowa/blob/main/docroot/profiles/custom/sitenow/tests/Kernel/PermissionsTest.php might offer some middle ground.

**Comment by richardbporter** [comment]
Noticed that the web container does not sync the `.git` directory with mutagen due to performance reasons. This means all Git operations need to happen on the host. Not much change there but I updated our BLT commands with Git ops to require being run on the host.

**Comment by richardbporter** [comment]
It seems like the uids_base Gulp watch task works well with ddev and the polling option is not necessary. I successfully watched and compiled SASS files with `ddev yarn workspace uids_base watch`.

**Comment by dnkemper** [comment]
Wanted to try this out and worked flawlessly 💅🏽

**Comment by joewhitsitt** [comment]
>Wanted to try this out and worked flawlessly 💅🏽

Let the record show your honor that we have successfully implemented DDEV on an old mac, a recent mac and a new M1.

**Review by dnkemper — APPROVED** [review]

**Comment by dnkemper** [comment]
Not sure if my approval is good enough for this, but this changed my life and I want to go back :)

**Comment by richardbporter** [comment]
Thanks but I forgot to test sitenow_migrate. 🤦🏻  Doing that now...

**Comment by joewhitsitt** [comment]
Probably want to get everyone currently on this before we merge into main.

**Comment by dnkemper** [comment]
> Probably want to get everyone currently on this before we merge into main.

FWIW I had to wipe everything out from testing this to get my vm environment back up and running..

**Comment by richardbporter** [comment]
> Thanks but I forgot to test sitenow_migrate. 🤦🏻 Doing that now...

Fixed.

**Comment by pyrello** [comment]
Testing this.

**Comment by pyrello** [comment]
Am I crazy or is there an extra `> uiowa:multisite:noop` at the end of `blt ds` commands?

**Comment by richardbporter** [comment]
> Am I crazy or is there an extra `> uiowa:multisite:noop` at the end of `blt ds` commands?

That was added in https://github.com/uiowa/uiowa/pull/4485. We have to specify a no-op for each command we want to override specified in BLT's default `sync.commands` configuration.

**Comment by pyrello** [comment]
Got it.

I am approving this with words since trying to get to the files page is causing my browser to timeout because of the number of file changes. Nice work!

**Comment by richardbporter** [comment]
Noticed the functional test written in 2595d9f15598f9b6896a0f58c9e2b280b2bbad83 isn't running in CI. Not a blocker, can be fixed in a follow up.

**Comment by joewhitsitt** [comment]
Converting back to draft until we are ready to merge. Need to update and account for any site provisions/deprovisions, etc. before merge

**Pull request merged**